pocs/camera/__init__.py

```diff
@@ -1,2 +1,157 @@
+from collections import OrderedDict
+
+from pocs.utils import error
+from pocs.utils import load_module
+from pocs.utils.config import load_config
+
 from pocs.camera.camera import AbstractCamera  # pragma: no flakes
 from pocs.camera.camera import AbstractGPhotoCamera  # pragma: no flakes
+
+from pocs.utils import list_connected_cameras
+from pocs.utils import logger as logger_module
+
+
+def create_cameras_from_config(config=None, logger=None, **kwargs):
+    """Creates a camera object(s)
+
+    Loads the cameras via the configuration.
+
+    Creates a camera for each camera item listed in the config. Ensures the
+    appropriate camera module is loaded.
+
+    Note:
+        This does not actually make a connection to the camera. To do so,
+        call 'camera.connect()' explicitly.
+
+    Args:
+        **kwargs (dict): Can pass a `cameras` object that overrides the info in
+            the configuration file. Can also pass `auto_detect`(bool) to try and
+            automatically discover the ports.
+
+    Returns:
+        OrderedDict: An ordered dictionary of created camera objects.
+
+    Raises:
+        error.CameraNotFound: Description
+        error.PanError: Description
+    """
+    if not logger:
+        logger = logger_module.get_root_logger()
+
+    if not config:
+        config = load_config(**kwargs)
+
+    if 'cameras' not in config:
+        logger.info('No camera information in config.')
+        return None
+
+    # Helper method to first check kwargs then config
+    def kwargs_or_config(item, default=None):
+        return kwargs.get(item, config.get(item, default))
+
+    camera_info = kwargs_or_config('cameras')
+    a_simulator = 'camera' in kwargs_or_config('simulator', default=list())
+    auto_detect = kwargs_or_config('auto_detect', default=False)
+
+    logger.debug("Camera config: {}".format(camera_info))
+
+    ports = list()
+
+    # Lookup the connected ports if not using a simulator
+    if not a_simulator and auto_detect:
+        logger.debug("Auto-detecting ports for cameras")
+        try:
+            ports = list_connected_cameras()
+        except Exception as e:
+            logger.warning(e)
+
+        if len(ports) == 0:
+            raise error.PanError(
+                msg="No cameras detected. Use --simulator=camera for simulator.")
+        else:
+            logger.debug("Detected Ports: {}".format(ports))
+
+    cameras = OrderedDict()
+    primary_camera = None
+
+    device_info = camera_info['devices']
+    for cam_num, device_config in enumerate(device_info):
+        cam_name = 'Cam{:02d}'.format(cam_num)
+
+        if not a_simulator:
+            camera_model = device_config.get('model')
+
+            # Assign an auto-detected port. If none are left, skip
+            if auto_detect:
+                try:
+                    camera_port = ports.pop()
```

Comment:
I'm perplexed by this. If the config file lists two canon cameras with the prefixes of their serial numbers, how do we know the order in which to pop the ports?

Reply:
This is only if auto_detect is true, which means that it is ignoring the config. Serial number is set when camera object is initialized by lookup (via gphoto2)

Reply:
Ah. Then I may have misunderstood how to auto_detect and primary work. I included the serial number prefixes and primary:True in pocs_local.yaml on PAN006. My goal is to make sure that a specific camera is treated as primary every time. And ideally we'd get a warning if a specified camera is not found, or if another camera is found.

Reply:
I think #95 is still true unfortunately. This function could certainly be cleaned up some. Happy to do it here if we want just to make sure it gets done soon but I imagine it will all need to be revisited anyway if we set up separate process controls for the camera. Which hopefully happens soon. :)